src/shaders/blend_atop_reverse/fragment.asm

```diff
@@ -0,0 +1,208 @@
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+pseq_to_dw_exec_nb = 13	// the number of 'EXEC' block where DW happens
+alu_buffer_size = 2	// number of .rgba regs carried through pipeline
+
+.uniforms
+	[5].l = "src_fmt_alpha";
+	[5].h = "src_swap_bgr";
+
+	[6].l = "mask_has_per_component_alpha";
+	[6].h = "mask_fmt_alpha";
+	[7].l = "mask_swap_bgr";
+	[7].h = "mask_clamp_to_border";
+
+	[8].l = "dst_fmt_alpha";
+	[8].h = "src_clamp_to_border";
+
+.asm
+
+// First batch
+EXEC
+	MFU:	sfu:  rcp r4
+		mul0: bar, sfu, bar0
+		mul1: bar, sfu, bar1
+		ipl:  NOP, NOP, t0.fp20, t0.fp20
+
+	// sample tex1 (mask)
+	TEX:	tex r0, r1, tex1, r2, r3, r0
+
+	// tmp = mask_fmt_alpha ? mask.a : 1.0
+	ALU:
+		ALU0:	CSEL lp.lh, -u6.h, r1.h, #1
+
+		// swap mask ABGR to ARGB if needed
+		ALU1:	CSEL lp.lh, -u7.l, r1.l, r0.l
+		ALU2:	CSEL lp.lh, -u7.l, r0.l, r1.l
+
+	// mask.r = mask_has_per_component_alpha ? mask.r : tmp
+	// mask.g = mask_has_per_component_alpha ? mask.g : tmp
+	// mask.b = mask_has_per_component_alpha ? mask.b : tmp
+	// mask.a = dst_fmt_alpha ? tmp : 0.0
+	ALU:
+		ALU0:	CSEL r0.l, -u6.l, alu1, alu0
+		ALU1:	CSEL r0.h, -u6.l, r0.h, alu0
+		ALU2:	CSEL r1.l, -u6.l, alu2, alu0
+		ALU3:	CSEL r1.h, -u8.l, alu0, #0
+;
+
+EXEC
+;
+
+// Second batch
+EXEC
+	// Emulate clamp-to-border for mask
+	ALU:
+		ALU0:	MAD  lp.lh, r2, #1, -#1
+		ALU1:	MAD  lp.lh, r3, #1, -#1
+
+	ALU:
+		ALU0:	CSEL lp.lh,   r2, u7.h, #0 (this)
+		ALU1:	CSEL lp.lh, alu0, #0, u7.h (other)
+		ALU2:	CSEL lp.lh,   r3, u7.h, #0 (other)
+		ALU3:	CSEL lp.lh, alu1, #0, u7.h
+
+	ALU:
+		ALU0:	MAD  r6.l, r0.l, #1, -alu0 (sat)
+		ALU1:	MAD  r6.h, r0.h, #1, -alu0 (sat)
+		ALU2:	MAD  r7.l, r1.l, #1, -alu0 (sat)
+		ALU3:	MAD  r7.h, r1.h, #1, -alu0 (sat)
+;
+
+EXEC
+;
+
+// Third batch
+EXEC
+	MFU:	sfu:  rcp r4
+		mul0: bar, sfu, bar0
+		mul1: bar, sfu, bar1
+		ipl:  t0.fp20, t0.fp20, NOP, NOP
+
+	// sample tex0 (src)
+	TEX:	tex r2, r3, tex0, r0, r1, r2
+
+	// src.a = src_fmt_alpha ? src.a : 1.0
+	ALU:
+		ALU0:	CSEL r5.h, -u5.l, r3.h, #1
+		ALU1:	MAD  r4.h, r2.h, #1, #0
+
+		// swap src ABGR to ARGB if needed
+		ALU2:	CSEL r4.l, -u5.h, r3.l, r2.l
+		ALU3:	CSEL r5.l, -u5.h, r2.l, r3.l
+;
+
+EXEC
+;
+
+// Fourth batch
+EXEC
+	// Emulate clamp-to-border for src
+	ALU:
+		ALU0:	MAD  lp.lh, r0, #1, -#1
+		ALU1:	MAD  lp.lh, r1, #1, -#1
+
+	ALU:
+		ALU0:	CSEL lp.lh,   r0, u8.h, #0 (this)
+		ALU1:	CSEL lp.lh, alu0, #0, u8.h (other)
+		ALU2:	CSEL lp.lh,   r1, u8.h, #0 (other)
+		ALU3:	CSEL lp.lh, alu1, #0, u8.h
+
+	ALU:
+		ALU0:	MAD  r4.l, r4.l, #1, -alu0 (sat)
+		ALU1:	MAD  r4.h, r4.h, #1, -alu0 (sat)
+		ALU2:	MAD  r5.l, r5.l, #1, -alu0 (sat)
+		ALU3:	MAD  r5.h, r5.h, #1, -alu0 (sat)
+;
+
+EXEC
+;
+
+// Fifth batch
+EXEC
+	// fetch dst pixel to r2,r3
+	PSEQ:	0x0081000A
```

Comment:
Right now we don't have any instructions for PSEQ in the asm. I think the instruction bits here should be: fetch enable, RT select and dst regs select. It would take me to write some test to verify this and to add `FETCH(?)` instruction to the fragment asm, I'll do it a bit later.